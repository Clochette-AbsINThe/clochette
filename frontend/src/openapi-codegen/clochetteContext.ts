import type { QueryKey, QueryOptions, UseQueryOptions } from '@tanstack/react-query';
import { useSession } from 'next-auth/react';
import { QueryOperation } from '../openapi-codegen/clochetteComponents';

export type ClochetteContext<
  TQueryFnData = unknown,
  TError = unknown,
  TData = TQueryFnData,
  TQueryKey extends QueryKey = QueryKey,
> = {
  fetcherOptions: {
    /**
     * Headers to inject in the fetcher
     */
    headers?: {
      authorization?: string;
    };
    /**
     * Query params to inject in the fetcher
     */
    queryParams?: Record<string, string | number | boolean>;
  };
  queryOptions: Omit<UseQueryOptions<TQueryFnData, TError, TData, TQueryKey>, 'queryKey' | "queryFn">;
  /**
   * Query key manager.
   */
  queryKeyFn: (operation: QueryOperation) => QueryKey;
};

/**
 * Context injected into every react-query hook wrappers
 *
 * @param queryOptions options from the useQuery wrapper
 */
export function useClochetteContext<TQueryFnData = unknown, TError = unknown, TData = TQueryFnData, TQueryKey extends QueryKey = QueryKey>(
  _queryOptions?: Omit<UseQueryOptions<TQueryFnData, TError, TData, TQueryKey>, 'queryKey' | 'queryFn'>
): ClochetteContext<TQueryFnData, TError, TData, TQueryKey> {
  const session = useSession();
  const token = session?.data?.token;
  return {
    fetcherOptions: {
      headers: {
        authorization: token ? `Bearer ${token}` : undefined,
      },
    },
    queryOptions: {
      retry: false,
      networkMode: 'always'
    },
    queryKeyFn
  };
}

export const queryKeyFn = (operation: QueryOperation) => {
  const queryKey: unknown[] = hasPathParams(operation)
    ? operation.path
      .split('/')
      .filter(Boolean)
      .map((i) => resolvePathParam(i, operation.variables.pathParams))
    : operation.path.split('/').filter(Boolean);

  if (hasQueryParams(operation)) {
    queryKey.push(operation.variables.queryParams);
  }

  if (hasBody(operation)) {
    queryKey.push(operation.variables.body);
  }

  return queryKey;
};
// Helpers
const resolvePathParam = (key: string, pathParams: Record<string, string>) => {
  if (key.startsWith('{') && key.endsWith('}')) {
    const slicedKey = key.slice(1, -1);
    const camelizeKey = slicedKey.replace(/_([a-z])/g, (g) => g[1].toUpperCase());
    return pathParams[camelizeKey];
  }
  return key;
};

const hasPathParams = (
  operation: QueryOperation
): operation is QueryOperation & {
  variables: { pathParams: Record<string, string> };
} => {
  return Boolean((operation.variables as any).pathParams);
};

const hasBody = (
  operation: QueryOperation
): operation is QueryOperation & {
  variables: { body: Record<string, unknown> };
} => {
  return Boolean((operation.variables as any).body);
};

const hasQueryParams = (
  operation: QueryOperation
): operation is QueryOperation & {
  variables: { queryParams: Record<string, unknown> };
} => {
  return Boolean((operation.variables as any).queryParams);
};
