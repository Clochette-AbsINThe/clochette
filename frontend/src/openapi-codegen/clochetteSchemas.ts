/**
 * Generated by @openapi-codegen
 *
 * @version 2.1.1
 */
/**
 * This this the account model that is linked to the database and used by the API.
 *
 * Args:
 *     AccountBase: The base model to use.
 */
export type Account = {
  /**
   * @maxLength 32
   * @minLength 3
   */
  username: string;
  lastName: string;
  firstName: string;
  /**
   * @maximum 2026
   * @minimum 2000
   */
  promotionYear: number;
  id: number;
  scope: SecurityScopes;
  isActive: boolean;
};

export type AccountCreate = {
  /**
   * @maxLength 32
   * @minLength 3
   */
  username: string;
  lastName: string;
  firstName: string;
  password: string;
  /**
   * @maximum 2026
   * @minimum 2000
   */
  promotionYear: number;
};

export type AccountUpdate = {
  username?: string | null;
  lastName?: string | null;
  firstName?: string | null;
  password?: string | null;
  promotionYear?: number | null;
  scope?: SecurityScopes | null;
  isActive?: boolean | null;
};

export type BarrelDistinct = {
  id: number;
  drinkItemId: number;
  buyPrice: number;
  sellPrice: number;
  barrelSellPrice: number | null;
  emptyOrSolded: boolean;
  isMounted: boolean;
  quantity: number;
  name: string;
};

export type BarrelUpdate = {
  fkId?: number | null;
  sellPrice?: number | null;
  unitPrice?: number | null;
  /**
   * @default false
   */
  empty?: boolean | null;
  /**
   * @default false
   */
  isMounted?: boolean | null;
};

export type BarrelUpdateModify = {
  sellPrice?: number | null;
  emptyOrSolded?: boolean | null;
  isMounted?: boolean | null;
};

export type BarrelUpdateSale = {
  /**
   * @exclusiveMinimum 0
   */
  barrelSellPrice: number;
  transactionId: number;
};

export type BodyLogin = {
  grant_type?: string | null;
  username: string;
  password: string;
  /**
   * @default
   */
  scope?: string;
  client_id?: string | null;
  client_secret?: string | null;
};

export type ConsumableDistinct = {
  id: number;
  consumableItemId: number;
  solded: boolean;
  sellPrice: number;
  buyPrice: number;
  quantity: number;
  name: string;
  icon: IconName;
};

export type ConsumableItem = {
  name: string;
  icon: IconName;
  id: number;
};

export type ConsumableItemCreate = {
  name: string;
  icon: IconName;
};

export type ConsumableItemUpdate = {
  name?: string | null;
  icon?: IconName | null;
};

export type ConsumableUpdate = {
  unitPrice?: number | null;
  sellPrice?: number | null;
};

export type ConsumableUpdateModify = {
  sellPrice?: number | null;
};

export type ConsumableUpdateSale = {
  /**
   * @default 0
   */
  transactionId?: number;
};

export type DrinkItem = {
  name: string;
  id: number;
};

export type DrinkItemCreate = {
  name: string;
};

export type DrinkItemUpdate = {
  name?: string | null;
};

export type HTTPError = {
  /**
   * Error message.
   */
  detail: string;
};

export type HTTPValidationError = {
  detail?: ValidationError[];
};

export type HealthResponse = {
  /**
   * OK
   */
  status: string;
};

export type IconName = 'Glass' | 'Beer' | 'Food' | 'Soft' | 'Barrel' | 'Misc';

export type Item = {
  table: string;
  /**
   * @exclusiveMinimum 0
   */
  quantity: number;
  item: Record<string, any> | ConsumableCreateV1 | OutOfStockCreate | BarrelCreateV1 | GlassCreateV1;
};

export type NonInventoried = {
  nonInventoriedItemId: number;
  buyPrice?: number | null;
  id: number;
  sellPrice: number | null;
  name: string;
  icon: IconName;
  trade: TradeType;
};

export type NonInventoriedCreate = {
  nonInventoriedItemId: number;
  buyPrice?: number | null;
  transactionId: number;
};

export type NonInventoriedItem = {
  /**
   * @minLength 1
   */
  name: string;
  icon: IconName;
  sellPrice?: number | null;
  id: number;
  trade: TradeType;
};

export type NonInventoriedItemCreate = {
  /**
   * @minLength 1
   */
  name: string;
  icon: IconName;
  sellPrice?: number | null;
};

export type NonInventoriedItemUpdate = {
  name?: string | null;
  icon?: IconName | null;
  sellPrice?: number | null;
};

export type OutOfStock = {
  fkId: number;
  unitPrice?: number | null;
  id: number;
  name: string;
  icon: IconName;
  sellPrice: number | null;
};

export type OutOfStockCreate = {
  fkId: number;
  unitPrice?: number | null;
  /**
   * @default 0
   */
  transaction_id?: number;
};

export type OutOfStockItem = {
  /**
   * @minLength 1
   */
  name: string;
  icon: IconName;
  sellPrice?: number | null;
  id: number;
  /**
   * buy_or_sell is True if the item does not have a sell price.
   */
  buyOrSell: boolean;
};

export type OutOfStockItemCreate = {
  /**
   * @minLength 1
   */
  name: string;
  icon: IconName;
  sellPrice?: number | null;
};

export type OutOfStockItemUpdate = {
  name?: string | null;
  icon?: IconName | null;
  sellPrice?: number | null;
};

export type OwnAccountUpdate = {
  username?: string | null;
  lastName?: string | null;
  firstName?: string | null;
  password?: string | null;
  promotionYear?: number | null;
};

export type PaymentMethod = 'CB' | 'Esp\u00E8ces' | 'Lydia' | 'Virement';

export type RootResponse = {
  /**
   * Hello, World!
   */
  msg: string;
};

export type SecurityScopes = 'staff' | 'treasurer' | 'president';

export type Status = 'pending' | 'validated';

export type Token = {
  access_token: string;
  /**
   * @default bearer
   */
  token_type?: string;
};

export type TradeType = 'purchase' | 'sale';

export type TransactionCommerceCreate = {
  /**
   * @format date-time
   */
  datetime: string;
  paymentMethod: PaymentMethod;
  trade: TradeType;
};

export type TransactionDetail = {
  /**
   * @format date-time
   */
  datetime: string;
  paymentMethod: PaymentMethod;
  trade: TradeType;
  id: number;
  treasuryId: number;
  type: TransactionType;
  status: Status;
  amount: number | null;
  description: string | null;
  barrelsPurchase: Barrel[];
  barrelsSale: Barrel[];
  glasses: Glass[];
  nonInventorieds: NonInventoried[];
  consumablesPurchase: Consumable[];
  consumablesSale: Consumable[];
};

export type TransactionFrontCreate = {
  /**
   * @format date-time
   */
  datetime: string;
  paymentMethod: PaymentMethod;
  sale: boolean;
  amount: number;
  /**
   * @default transaction
   */
  type?: TransactionTypeV1;
  description?: string | null;
  items?: Item[];
};

export type TransactionSingle = {
  /**
   * @format date-time
   */
  datetime: string;
  paymentMethod: PaymentMethod;
  sale: boolean;
  amount: number;
  /**
   * @default transaction
   */
  type?: TransactionTypeV1;
  description?: string | null;
  id: number;
  treasuryId: number;
  barrels: BarrelV1[] | null;
  glasses: GlassV1[] | null;
  outOfStocks: OutOfStock[] | null;
  consumablesPurchase: ConsumableV1[] | null;
  consumablesSale: ConsumableV1[] | null;
};

export type TransactionTreasuryCreate = {
  /**
   * @format date-time
   */
  datetime: string;
  paymentMethod: PaymentMethod;
  trade: TradeType;
  amount: number;
  description: string;
};

export type TransactionType = 'commerce' | 'tresorery';

export type TransactionTypeV1 = 'transaction' | 'tresorery';

export type Treasury = {
  totalAmount: number;
  /**
   * @minimum 0
   */
  cashAmount: number;
  /**
   * @maximum 1
   * @minimum 0
   */
  lydiaRate: number;
  id: number;
};

export type TreasuryUpdate = {
  lydiaRate?: number | null;
};

export type ValidationError = {
  loc: (string | number)[];
  msg: string;
  type: string;
};

export type VersionResponse = {
  /**
   * Version of the API.
   */
  version: string;
};

export type BarrelV1 = {
  fkId?: number | null;
  /**
   * @exclusiveMinimum 0
   */
  sellPrice: number;
  /**
   * @exclusiveMinimum 0
   */
  unitPrice: number;
  id: number;
  empty: boolean;
  isMounted: boolean;
  name: string;
};

export type BarrelCreateV1 = {
  fkId?: number | null;
  /**
   * @exclusiveMinimum 0
   */
  sellPrice: number;
  /**
   * @exclusiveMinimum 0
   */
  unitPrice: number;
  /**
   * @default 0
   */
  transaction_id?: number;
};

export type ConsumableV1 = {
  /**
   * @exclusiveMinimum 0
   */
  unitPrice: number;
  /**
   * @exclusiveMinimum 0
   */
  sellPrice: number;
  id: number;
  fkId: number;
  empty: boolean;
  name: string;
  icon: IconName;
};

export type ConsumableCreateV1 = {
  /**
   * @exclusiveMinimum 0
   */
  unitPrice: number;
  /**
   * @exclusiveMinimum 0
   */
  sellPrice: number;
  fkId: number;
  id?: number | null;
};

export type GlassV1 = {
  fkId: number;
  id: number;
  name: string;
  sellPrice: number;
};

export type GlassCreateV1 = {
  fkId: number;
  /**
   * @default 0
   */
  transaction_id?: number;
};

export type TransactionV1 = {
  /**
   * @format date-time
   */
  datetime: string;
  paymentMethod: PaymentMethod;
  sale: boolean;
  amount: number;
  /**
   * @default transaction
   */
  type?: TransactionTypeV1;
  description?: string | null;
  id: number;
  treasuryId: number;
};

export type Barrel = {
  id: number;
  drinkItemId: number;
  buyPrice: number;
  sellPrice: number;
  barrelSellPrice: number | null;
  emptyOrSolded: boolean;
  isMounted: boolean;
  name: string;
};

export type BarrelCreate = {
  drinkItemId: number;
  /**
   * @exclusiveMinimum 0
   */
  sellPrice: number;
  /**
   * @exclusiveMinimum 0
   */
  buyPrice: number;
  transactionId: number;
};

export type Consumable = {
  id: number;
  consumableItemId: number;
  solded: boolean;
  sellPrice: number;
  buyPrice: number;
  name: string;
  icon: IconName;
};

export type ConsumableCreate = {
  consumableItemId: number;
  /**
   * @exclusiveMinimum 0
   */
  sellPrice: number;
  /**
   * @exclusiveMinimum 0
   */
  buyPrice: number;
  /**
   * @default 0
   */
  transactionId?: number;
};

export type Glass = {
  barrelId: number;
  id: number;
  sellPrice: number;
  name: string;
};

export type GlassCreate = {
  barrelId: number;
  transactionId: number;
};

export type Transaction = {
  /**
   * @format date-time
   */
  datetime: string;
  paymentMethod: PaymentMethod;
  trade: TradeType;
  id: number;
  treasuryId: number;
  type: TransactionType;
  status: Status;
  amount: number | null;
  description: string | null;
};
