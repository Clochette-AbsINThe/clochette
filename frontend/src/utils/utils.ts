import { AxiosResponse } from 'axios';
import { environmentVariable } from '@utils/settings';

export const addIdToUrl = (id: number): void => {
    const url = new URL(window.location.href);
    url.searchParams.set('id', id.toString());
    window.history.pushState({}, '', url.href);
};

export const removeIdFromUrl = (): void => {
    const url = new URL(window.location.href);
    url.searchParams.delete('id');
    window.history.pushState({}, '', url.href);
};

export const getIdFromUrl = (): number | null => {
    const url = new URL(window.location.href);
    const id = url.searchParams.get('id');
    if (id === null) return null;
    if (/^(-|)[0-9]+$/.test(id)) return parseInt(id);
    return null;
};

export function getErrorMessage(data: AxiosResponse<unknown, any>): string {
    if (data.status === 422) {
        const { detail } = data.data as {
            detail: [
                {
                    loc: [string, number];
                    msg: string;
                    type: string;
                }
            ];
        };
        return (
            detail
                .map((e) => e.msg)
                .join(' ')
                .charAt(0)
                .toUpperCase() +
            detail
                .map((e) => e.msg)
                .join(' ')
                .slice(1)
        );
    } else if (data.status === 500) {
        return 'Erreur serveur';
    } else {
        const { detail } = data.data as { detail: string };
        return detail;
    }
}

export interface Token {
    sub: string;
    iat: number;
    exp: number;
    token_type: string;
}

/**
 * This function could throw an error if the token is invalid in its format, for example if the token is not a JWT token
 * @param token JWT token
 * @returns Inforamtions inside the token
 */
export function parseJwt(token: string): Token {
    const base64Url = token.split('.')[1];
    const base64 = base64Url?.replace(/-/g, '+').replace(/_/g, '/');
    const buffer = Buffer.from(base64 as string, 'base64');
    const jsonPayload = decodeURIComponent(
        buffer
            .toString('utf-8')
            .split('')
            .map((c) => '%' + ('00' + c.charCodeAt(0).toString(16)).slice(-2))
            .join('')
    );
    return JSON.parse(jsonPayload);
}

export function getRedirectUrlEncoded(url: string): string {
    return encodeURIComponent(environmentVariable.BASE_URL + url);
}
