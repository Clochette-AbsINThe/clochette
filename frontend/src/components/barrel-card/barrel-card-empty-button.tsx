import { useState } from 'react';
import toast from 'react-hot-toast';

import { DialogClose } from '@radix-ui/react-dialog';
import { useQueryClient } from '@tanstack/react-query';

import { Button as QueryButton } from '@/components/button';
import { Button } from '@/components/ui/button';
import { Dialog, DialogContent, DialogDescription, DialogFooter, DialogHeader, DialogTitle, DialogTrigger } from '@/components/ui/dialog';
import { useUpdateBarrel } from '@/openapi-codegen/clochetteComponents';
import { generateApiErrorMessage } from '@/openapi-codegen/clochetteFetcher';
import { Barrel } from '@/openapi-codegen/clochetteSchemas';

export function EmptyBarrelButton({ barrel }: Readonly<{ barrel: Barrel }>) {
  const queryClient = useQueryClient();
  const [isOpen, setIsOpen] = useState(false);

  const modifyBarrel = useUpdateBarrel({
    onSuccess(data, variables, context) {
      toast.success(`Fût ${data.name} vidé avec succès !`);
      queryClient.invalidateQueries({ stale: true });
      setIsOpen(false);
    },
    onError(error, variables, context) {
      const detail = generateApiErrorMessage(error);
      toast.error(`Erreur lors de la modification du fût. ${detail}`);
    }
  });

  const onClick = () => {
    modifyBarrel.mutate({
      pathParams: {
        barrelId: barrel.id
      },
      body: {
        isMounted: false,
        emptyOrSolded: true
      }
    });
  };

  return (
    <Dialog
      open={isOpen}
      onOpenChange={setIsOpen}
    >
      <DialogTrigger asChild>
        <Button
          variant={'destructive'}
          className='w-full'
        >
          Fût vide
        </Button>
      </DialogTrigger>
      <DialogContent>
        <DialogHeader>
          <DialogTitle>Vider le fût</DialogTitle>
          <DialogDescription>Etes-vous sûr de vouloir déclarer le fût vide ?</DialogDescription>
        </DialogHeader>
        <DialogFooter>
          <div className='flex justify-end space-x-5'>
            <DialogClose asChild>
              <QueryButton>Annuler</QueryButton>
            </DialogClose>
            <QueryButton
              danger
              loading={modifyBarrel.isPending}
              onClick={onClick}
            >
              Vider
            </QueryButton>
          </div>
        </DialogFooter>
      </DialogContent>
    </Dialog>
  );
}
