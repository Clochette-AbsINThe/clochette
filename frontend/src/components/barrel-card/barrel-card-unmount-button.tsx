import toast from 'react-hot-toast';

import { Cross2Icon, UpdateIcon } from '@radix-ui/react-icons';
import { useQueryClient } from '@tanstack/react-query';

import { Tooltip, TooltipContent, TooltipProvider, TooltipTrigger } from '@/components/ui/tooltip';
import { useUpdateBarrel } from '@/openapi-codegen/clochetteComponents';
import { generateApiErrorMessage } from '@/openapi-codegen/clochetteFetcher';
import { Barrel } from '@/openapi-codegen/clochetteSchemas';

export function UnmountBarrelButton({ barrel }: { barrel: Barrel }) {
  const queryClient = useQueryClient();

  const modifyBarrel = useUpdateBarrel({
    onSuccess(data, variables, context) {
      toast.success(`Fût ${data.name} démonté avec succès !`);
      queryClient.invalidateQueries({ stale: true });
    },
    onError(error, variables, context) {
      const detail = generateApiErrorMessage(error);
      toast.error(`Erreur lors de la modification du fût. ${detail}`);
    }
  });

  const onClick = () => {
    modifyBarrel.mutate({
      pathParams: {
        barrelId: barrel.id
      },
      body: {
        isMounted: false
      }
    });
  };

  return (
    <TooltipProvider>
      <Tooltip>
        <TooltipTrigger asChild>
          <button
            disabled={modifyBarrel.isPending}
            onClick={onClick}
            className='absolute right-4 top-4 rounded-sm opacity-70 ring-offset-background transition-opacity hover:opacity-100 focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 disabled:pointer-events-none data-[state=open]:bg-accent data-[state=open]:text-muted-foreground'
          >
            {modifyBarrel.isPending ? <UpdateIcon className='h-4 w-4 animate-spin' /> : <Cross2Icon className='h-4 w-4' />}
            <span className='sr-only'>Unmount</span>
          </button>
        </TooltipTrigger>
        <TooltipContent>
          <p>Remettre le fût en stock</p>
        </TooltipContent>
      </Tooltip>
    </TooltipProvider>
  );
}
