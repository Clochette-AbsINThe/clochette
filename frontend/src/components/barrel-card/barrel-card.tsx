import { useState } from 'react';
import toast from 'react-hot-toast';

import { DialogClose } from '@radix-ui/react-dialog';
import { Cross2Icon, UpdateIcon } from '@radix-ui/react-icons';
import { useQueryClient } from '@tanstack/react-query';

import { Button as QueryButton } from '@/components/button';
import { Button } from '@/components/ui/button';
import { Dialog, DialogContent, DialogDescription, DialogFooter, DialogHeader, DialogTitle, DialogTrigger } from '@/components/ui/dialog';
import { Tooltip, TooltipContent, TooltipProvider, TooltipTrigger } from '@/components/ui/tooltip';
import { useReadGlasses, useUpdateBarrel } from '@/openapi-codegen/clochetteComponents';
import { generateApiErrorMessage } from '@/openapi-codegen/clochetteFetcher';
import { Barrel } from '@/openapi-codegen/clochetteSchemas';
import { getIcon } from '@/styles/utils';

export function UnmountBarrelButton({ barrel }: { barrel: Barrel }) {
  const queryClient = useQueryClient();

  const modifyBarrel = useUpdateBarrel({
    onSuccess(data, variables, context) {
      toast.success(`Fût ${data.name} démonté avec succès !`);
      queryClient.invalidateQueries({ stale: true });
    },
    onError(error, variables, context) {
      const detail = generateApiErrorMessage(error);
      toast.error(`Erreur lors de la modification du fût. ${detail}`);
    }
  });

  const onClick = () => {
    modifyBarrel.mutate({
      pathParams: {
        barrelId: barrel.id
      },
      body: {
        isMounted: false
      }
    });
  };

  return (
    <TooltipProvider>
      <Tooltip>
        <TooltipTrigger asChild>
          <button
            disabled={modifyBarrel.isLoading}
            onClick={onClick}
            className='absolute right-4 top-4 rounded-sm opacity-70 ring-offset-background transition-opacity hover:opacity-100 focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 disabled:pointer-events-none data-[state=open]:bg-accent data-[state=open]:text-muted-foreground'
          >
            {modifyBarrel.isLoading ? <UpdateIcon className='h-4 w-4 animate-spin' /> : <Cross2Icon className='h-4 w-4' />}
            <span className='sr-only'>Unmount</span>
          </button>
        </TooltipTrigger>
        <TooltipContent>
          <p>Remettre le fût en stock</p>
        </TooltipContent>
      </Tooltip>
    </TooltipProvider>
  );
}

export function EmptyBarrelButton({ barrel }: { barrel: Barrel }) {
  const queryClient = useQueryClient();
  const [isOpen, setIsOpen] = useState(false);

  const modifyBarrel = useUpdateBarrel({
    onSuccess(data, variables, context) {
      toast.success(`Fût ${data.name} vidé avec succès !`);
      queryClient.invalidateQueries({ stale: true });
      setIsOpen(false);
    },
    onError(error, variables, context) {
      const detail = generateApiErrorMessage(error);
      toast.error(`Erreur lors de la modification du fût. ${detail}`);
    }
  });

  const onClick = () => {
    modifyBarrel.mutate({
      pathParams: {
        barrelId: barrel.id
      },
      body: {
        isMounted: false,
        emptyOrSolded: true
      }
    });
  };

  return (
    <Dialog
      open={isOpen}
      onOpenChange={setIsOpen}
    >
      <DialogTrigger asChild>
        <Button
          variant={'destructive'}
          className='w-full'
        >
          Fût vide
        </Button>
      </DialogTrigger>
      <DialogContent>
        <DialogHeader>
          <DialogTitle>Vider le fût</DialogTitle>
          <DialogDescription>Etes-vous sûr de vouloir déclarer le fût vide ?</DialogDescription>
        </DialogHeader>
        <DialogFooter>
          <div className='flex justify-end space-x-5'>
            <DialogClose asChild>
              <QueryButton>Annuler</QueryButton>
            </DialogClose>
            <QueryButton
              danger
              loading={modifyBarrel.isLoading}
              onClick={onClick}
            >
              Vider
            </QueryButton>
          </div>
        </DialogFooter>
      </DialogContent>
    </Dialog>
  );
}

export function BarrelCard({ barrel }: { barrel: Barrel }) {
  const { data } = useReadGlasses({
    queryParams: {
      barrel_id: barrel.id
    }
  });

  return (
    <div className='flex bg-card rounded-lg shadow-md border relative px-6'>
      <UnmountBarrelButton barrel={barrel} />
      <div className='flex flex-col px-2 py-4 items-center w-full gap-2 mt-2'>
        <h1 className='text-xl font-semibold uppercase [hyphens:_auto] text-center'>{barrel.name}</h1>
        {getIcon('Barrel', 'h-24 w-24 text-gray-500 dark:text-gray-100')}
        <ul className='list-disc pl-6 mb-2'>
          <li className='text-md text-muted-foreground'>Prix de vente au verre: {barrel.sellPrice}€</li>
          {data && <li className='text-md text-muted-foreground'>Verres vendus: {data.length}</li>}
        </ul>
        <EmptyBarrelButton barrel={barrel} />
      </div>
    </div>
  );
}
