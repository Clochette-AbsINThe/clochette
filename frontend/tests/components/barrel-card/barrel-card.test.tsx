import { render, screen } from '@testing-library/react';

import { BarrelCard } from '@/components/barrel-card/barrel-card';
import { Barrel } from '@/openapi-codegen/clochetteSchemas';
import { createWrapper } from 'tests/utils';

const mockBarrel: Barrel = {
  id: 1,
  name: 'Mock Barrel',
  sellPrice: 5.0
} as any;

describe('BarrelCard', () => {
  it('renders the barrel card', () => {
    const { container } = render(<BarrelCard barrel={mockBarrel} />, {
      wrapper: createWrapper()
    });
    expect(container).toMatchSnapshot();
    expect(screen.getByText('Mock Barrel')).not.toBeNull();
  });
});
