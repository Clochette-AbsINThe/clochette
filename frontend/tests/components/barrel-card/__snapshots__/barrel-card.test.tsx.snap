// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BarrelCard > renders the barrel card 1`] = `
<div>
  <div
    class="flex bg-card rounded-lg shadow-md border relative px-6"
  >
    <button
      class="absolute right-4 top-4 rounded-sm opacity-70 ring-offset-background transition-opacity hover:opacity-100 focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 disabled:pointer-events-none data-[state=open]:bg-accent data-[state=open]:text-muted-foreground"
      data-state="closed"
    >
      <svg
        class="h-4 w-4"
        fill="none"
        height="15"
        viewBox="0 0 15 15"
        width="15"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M11.7816 4.03157C12.0062 3.80702 12.0062 3.44295 11.7816 3.2184C11.5571 2.99385 11.193 2.99385 10.9685 3.2184L7.50005 6.68682L4.03164 3.2184C3.80708 2.99385 3.44301 2.99385 3.21846 3.2184C2.99391 3.44295 2.99391 3.80702 3.21846 4.03157L6.68688 7.49999L3.21846 10.9684C2.99391 11.193 2.99391 11.557 3.21846 11.7816C3.44301 12.0061 3.80708 12.0061 4.03164 11.7816L7.50005 8.31316L10.9685 11.7816C11.193 12.0061 11.5571 12.0061 11.7816 11.7816C12.0062 11.557 12.0062 11.193 11.7816 10.9684L8.31322 7.49999L11.7816 4.03157Z"
          fill="currentColor"
          fill-rule="evenodd"
        />
      </svg>
      <span
        class="sr-only"
      >
        Unmount
      </span>
    </button>
    <div
      class="flex flex-col px-2 py-4 items-center w-full gap-2 mt-2"
    >
      <h1
        class="text-xl font-semibold uppercase [hyphens:_auto] text-center"
      >
        Mock Barrel
      </h1>
      <svg
        aria-label="icon"
        class="h-24 w-24 text-gray-500 dark:text-gray-100"
        data-testid="Barrel"
        viewBox="0 0 24 24"
      >
        <g
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="0.5"
        >
          <path
            d="M7.278 4h9.444a2 2 0 0 1 1.841 1.22C19.521 7.48 20 9.74 20 12c0 2.26-.479 4.52-1.437 6.78A2 2 0 0 1 16.722 20H7.278a2 2 0 0 1-1.841-1.22C4.479 16.52 4 14.26 4 12c0-2.26.479-4.52 1.437-6.78A2 2 0 0 1 7.278 4z"
          />
          <path
            d="M14 4c.667 2.667 1 5.333 1 8s-.333 5.333-1 8M10 4c-.667 2.667-1 5.333-1 8s.333 5.333 1 8m-5.5-4h15m0-8h-15"
          />
        </g>
      </svg>
      <ul
        class="list-disc pl-6 mb-2"
      >
        <li
          class="text-md text-muted-foreground"
        >
          Prix de vente au verre: 
          5
          €
        </li>
      </ul>
      <button
        aria-controls="radix-:r1:"
        aria-expanded="false"
        aria-haspopup="dialog"
        class="inline-flex items-center justify-center rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-destructive text-destructive-foreground shadow-sm hover:bg-destructive/90 h-9 px-4 py-2 w-full"
        data-state="closed"
        type="button"
      >
        Fût vide
      </button>
    </div>
  </div>
</div>
`;
